proxy/handlers/handlers.go

```diff
@@ -0,0 +1,171 @@
+		if err != nil {
+			proxy.Logger.Debugf("error converting metric: %v", err)
+			errorCount += 1
+			continue
+		}
+		err = proxy.handleMetric(metric)
+		if err != nil {
+			errorCount += 1
+		}
+	}
+
+	proxy.Statsd.Count(
+		"veneur_proxy.ingest.metrics_count", int64(len(jsonMetrics)-errorCount),
+		[]string{"error:false", "protocol:http"}, 1.0)
+	if errorCount > 0 {
+		proxy.Statsd.Count(
+			"veneur_proxy.ingest.metrics_count", int64(errorCount),
+			[]string{"error:true", "protocol:http"}, 1.0)
+	}
+}
+
+// Receives and handles metrics via individual gRPC requests.
+func (proxy *Handlers) SendMetrics(
+	ctx context.Context, metricList *forwardrpc.MetricList,
+) (*empty.Empty, error) {
+	proxy.Statsd.Count(
+		"veneur_proxy.ingest.request_count", 1,
+		[]string{"protocol:grpc-single"}, 1.0)
+
+	errorCount := 0
+	for _, metric := range metricList.Metrics {
+		err := proxy.handleMetric(metric)
+		if err != nil {
+			errorCount += 1
+		}
+	}
+
+	proxy.Statsd.Count(
+		"veneur_proxy.ingest.metrics_count",
+		int64(len(metricList.Metrics)-errorCount),
+		[]string{"error:false", "protocol:grpc-single"}, 1.0)
+	if errorCount > 0 {
+		proxy.Statsd.Count(
+			"veneur_proxy.ingest.metrics_count", int64(errorCount),
+			[]string{"error:true", "protocol:grpc-single"}, 1.0)
+	}
+	return &emptypb.Empty{}, nil
+}
+
+// Receives and handles metrics via gRPC streaming.
+func (proxy *Handlers) SendMetricsV2(
+	server forwardrpc.Forward_SendMetricsV2Server,
+) error {
+	proxy.Statsd.Count(
+		"veneur_proxy.ingest.request_count", 1,
+		[]string{"protocol:grpc-stream"}, 1.0)
+
+	for {
+		metric, err := server.Recv()
+		if err == io.EOF {
+			return nil
+		} else if err != nil {
+			proxy.Logger.WithError(err).Debug("error receiving metrics")
+			proxy.Statsd.Count(
+				"veneur_proxy.ingest.request_error_count", 1,
+				[]string{"protocol:grpc-stream"}, 1.0)
+			return err
+		}
+		err = proxy.handleMetric(metric)
+		if err != nil {
+			proxy.Statsd.Count(
+				"veneur_proxy.ingest.metrics_count",
+				int64(1), []string{"error:true", "protocol:grpc-stream"}, 1.0)
+		} else {
+			proxy.Statsd.Count(
+				"veneur_proxy.ingest.metrics_count",
+				int64(1), []string{"error:false", "protocol:grpc-stream"}, 1.0)
+		}
+	}
+}
+
+// Forwards metrics to downstream global Veneur instances.
+func (proxy *Handlers) handleMetric(metric *metricpb.Metric) error {
+	tags := []string{}
+tagLoop:
+	for _, tag := range metric.Tags {
+		for _, matcher := range proxy.IgnoreTags {
+			if matcher.Match(tag) {
+				continue tagLoop
+			}
+		}
+		tags = append(tags, tag)
+	}
+
+	key := fmt.Sprintf("%s%s%s",
+		metric.Name, strings.ToLower(metric.Type.String()), strings.Join(tags, ","))
+	destination, err := proxy.Destinations.Get(key)
+	if err != nil {
+		proxy.Logger.WithError(err).Debug("failed to get destination")
+		proxy.Statsd.Count(
+			"veneur_proxy.forward.metrics_count", 1,
+			[]string{"error:true"}, 1.0)
+		return err
+	}
+
+	errorChannel := make(chan error)
```

Comment:
1. I think this callers of this handleMetric method just use error count rather than content of error
2. where is the close channel statement? is it needed?

Reply:
1. that is true; but an error (nil or non-nill pointer) isn't structurally different from a boolean (zero or one).
2. ah good catch, that does look like a leak to me. I will fix. Thanks!

Reply:
Actually, looking at docs (https://stackoverflow.com/a/36613932) it seems like channels are correctly garbage collected.

Reply:
Also, I just realized, this channel is closed in connect.go:
```
close(request.ErrorChannel)
```